src/apps/vguest/vguest_dev.lua

```diff
@@ -0,0 +1,227 @@
+   local fd, err = S.open(fname, 'rdwr')
+   if not fd then error(err) end
+   return setmetatable ({
+      fd = fd,
+      struct = struct,
+      close = function(self) return self.fd:close() end,
+   }, {
+      __index = function (self, key)
+         return fieldrd(self.struct[key], self.fd)
+      end,
+      __newindex = function (self, key, value)
+         return fieldwr(self.struct[key], self.fd, value)
+      end,
+   })
+end
+
+virtio_pci_bar0 = fstruct[[
+   uint32_t host_features;
+   uint32_t guest_features;
+   uint32_t queue_pfn;
+   uint16_t queue_num;
+   uint16_t queue_sel;
+   uint16_t queue_notify;
+   uint8_t status;
+   uint8_t isr;
+   uint16_t config_vector;
+   uint16_t queue_vector;
+]]
+
+
+VGdev = {}
+VGdev.__index = VGdev
+
+
+function VGdev:new(args)
+   local min_features = 0 -- C.VIRTIO_F_VERSION_1?
+   local want_features = C.VIRTIO_NET_F_CSUM
+                        + C.VIRTIO_NET_F_MAC
+--                         + C.VIRTIO_F_VERSION_1
+   pci.unbind_device_from_linux (args.pciaddr)
+
+   local bar = openBar(pci.path(args.pciaddr..'/resource0'), virtio_pci_bar0)
+--    for k,v in pairs(virtio_pci_bar0) do
+--       print (string.format('%s: %X', v.fieldname, bar[v.fieldname]))
+--    end
+
+   bar.status = 0           -- reset device
+   bar.status = bit.bor(bar.status, 1)           -- acknowledge
+   -- check something
+   bar.status = bit.bor(bar.status, 2)           -- driver
+   local features = bar.host_features
+   print ('host_features', features)
+   if bit.band(features, min_features) ~= min_features then
+      bar.status = bit.bor(bar.status, 128)      -- failure
+      bar:close()
+      return nil, "doesn't provide minimum features"
+   end
+   print ('set features to:', bit.band(features, want_features))
+   bar.guest_features = bit.band(features, want_features)
+   bar.status = bit.bor(bar.status, 8)           -- features_ok
+   print ('got features: ', bar.host_features, bar.guest_features)
+   if bit.band(bar.status, 8) ~= 8 then
+      bar.status = bit.bor(bar.status, 128)      -- failure
+      bar:close()
+      return nil, "feature set wasn't accepted by device"
+   end
+
+   print ("enumerating queues...")
+   local vqs = {}
+   for qn = 0, 16 do
+      bar.queue_sel = qn
+      local queue_size = bar.queue_num
+      if queue_size == 0 then break end
+
+      print (string.format('queue %d: size: %d', qn, queue_size))
+      local vring = gvring.allocate_vring(bar.queue_num)
+      vqs[qn] = vring
+      bar.queue_pfn = bit.rshift(vring.vring_physaddr, 12)      -- VIRTIO_PCI_QUEUE_ADDR_SHIFT
+   end
+
+   if not(vqs[0] and vqs[1]) then
+      bar.status = bit.bor(bar.status, 128)      -- failure
+      bar:close()
+      return nil, "missing required virtqueues"
+   end
+
+   bar.status = bit.bor(bar.status, 4)           -- driver_ok
+
+   return setmetatable({
+      bar = bar,
+      vqs = vqs,
+   }, self)
+end
+
+
+function VGdev:close()
+   for qn, vq in pairs(self.vqs) do
```

Comment:
`vq` is not used, should be assigned to underscore. In addition to that, I think `self.vqs` is an array, if that's the case it's better to use `ipairs` as `pairs` is much slower.